File: dotnet/src/Envdoctor/Utils/Locale.cs
namespace Envdoctor.Utils;

/// A comparator matching JavaScript's `String.prototype.localeCompare` for the
/// character set used in environment-variable names (`[A-Za-z0-9_]`).
///
/// The TypeScript reference sorts generated output with `localeCompare`, which
/// is a Unicode collation (punctuation < digits < letters, case-insensitive
/// with a lowercase-before-uppercase tiebreak) — not byte order.
public static class Locale
{
    private static int Primary(char c)
    {
        if (char.IsAsciiDigit(c))
            return 1000 + c;
        if (char.IsAsciiLetter(c))
            return 2000 + char.ToUpperInvariant(c);
        return c;
    }

    private static int CaseWeight(char c) => char.IsAsciiLetterLower(c) ? 0 : 1;

    /// Compare two strings the way JS `a.localeCompare(b)` does for env-var names.
    public static int LocaleCompare(string a, string b)
    {
        var len = Math.Min(a.Length, b.Length);
        // Level 1: primary weights across all positions.
        for (var i = 0; i < len; i++)
        {
            var cmp = Primary(a[i]).CompareTo(Primary(b[i]));
            if (cmp != 0)
                return cmp;
        }
        var lenCmp = a.Length.CompareTo(b.Length);
        if (lenCmp != 0)
            return lenCmp;

        // Level 2: case (only reached when primaries are all equal).
        for (var i = 0; i < len; i++)
        {
            var cmp = CaseWeight(a[i]).CompareTo(CaseWeight(b[i]));
            if (cmp != 0)
                return cmp;
        }
        return 0;
    }

    public static void LocaleSort(List<string> values) => values.Sort(LocaleCompare);
}

File: kotlin/src/main/kotlin/io/github/arun_skg/envdoctor/utils/Locale.kt
package io.github.arun_skg.envdoctor.utils

/**
 * A comparator matching JavaScript's `String.prototype.localeCompare` for the
 * character set used in environment-variable names (`[A-Za-z0-9_]`).
 *
 * The TypeScript reference sorts generated output with `localeCompare`, which
 * is a Unicode collation (punctuation < digits < letters, case-insensitive
 * with a lowercase-before-uppercase tiebreak) — not byte order.
 */
object Locale {
    private fun primary(c: Char): Int = when {
        c in '0'..'9' -> 1000 + c.code
        c.isAsciiLetter() -> 2000 + c.uppercaseChar().code
        else -> c.code
    }

    private fun caseWeight(c: Char): Int = if (c in 'a'..'z') 0 else 1

    private fun Char.isAsciiLetter(): Boolean = this in 'a'..'z' || this in 'A'..'Z'

    /** Compare two strings the way JS `a.localeCompare(b)` does for env-var names. */
    fun localeCompare(a: String, b: String): Int {
        val len = minOf(a.length, b.length)
        // Level 1: primary weights across all positions.
        for (i in 0 until len) {
            val cmp = primary(a[i]) - primary(b[i])
            if (cmp != 0) return cmp
        }
        val lenCmp = a.length - b.length
        if (lenCmp != 0) return lenCmp

        // Level 2: case (only reached when primaries are all equal).
        for (i in 0 until len) {
            val cmp = caseWeight(a[i]) - caseWeight(b[i])
            if (cmp != 0) return cmp
        }
        return 0
    }

    fun localeSort(values: MutableList<String>) {
        values.sortWith(::localeCompare)
    }
}
